core/trino-main/src/test/java/io/trino/sql/planner/optimizations/TestAddLocalExchangesForTaskScaleWriters.java

```diff
@@ -0,0 +1,110 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.sql.planner.optimizations;
+
+import com.google.common.collect.ImmutableMap;
+import io.trino.Session;
+import io.trino.connector.MockConnectorFactory;
+import io.trino.plugin.tpch.TpchConnectorFactory;
+import io.trino.sql.planner.LogicalPlanner;
+import io.trino.sql.planner.SubPlan;
+import io.trino.sql.planner.assertions.BasePlanTest;
+import io.trino.sql.planner.plan.ExchangeNode;
+import io.trino.testing.LocalQueryRunner;
+import org.intellij.lang.annotations.Language;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import static io.trino.SystemSessionProperties.SCALE_WRITERS;
+import static io.trino.SystemSessionProperties.TASK_SCALE_WRITERS_ENABLED;
+import static io.trino.sql.planner.SystemPartitioningHandle.SCALED_WRITER_DISTRIBUTION;
+import static io.trino.sql.planner.plan.ExchangeNode.Scope.LOCAL;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestAddLocalExchangesForTaskScaleWriters
+        extends BasePlanTest
+{
+    @Override
+    protected LocalQueryRunner createLocalQueryRunner()
+    {
+        Session session = testSessionBuilder()
+                .setCatalog("tpch")
+                .setSchema("tiny")
+                .build();
+        LocalQueryRunner queryRunner = LocalQueryRunner.create(session);
+        queryRunner.createCatalog("tpch", new TpchConnectorFactory(1), ImmutableMap.of());
+        queryRunner.createCatalog("mock_dont_report_written_bytes", createConnectorFactorySupportingReportingBytesWritten(false, "mock_dont_report_written_bytes"), ImmutableMap.of());
+        queryRunner.createCatalog("mock_report_written_bytes", createConnectorFactorySupportingReportingBytesWritten(true, "mock_report_written_bytes"), ImmutableMap.of());
+        return queryRunner;
+    }
+
+    private MockConnectorFactory createConnectorFactorySupportingReportingBytesWritten(boolean supportsWrittenBytes, String name)
+    {
+        return MockConnectorFactory.builder()
+                .withSupportsReportingWrittenBytes(supportsWrittenBytes)
+                .withGetTableHandle(((session, schemaTableName) -> null))
+                .withName(name)
+                .build();
+    }
+
+    @DataProvider
+    public Object[][] taskScaleWritersEnabled()
+    {
+        return new Object[][] {{true}, {false}};
+    }
+
+    @Test(dataProvider = "taskScaleWritersEnabled")
+    public void testLocalScaledWriterDistributionWithSupportsReportingWrittenBytes(boolean isTaskScaleWritersEnabled)
```

Comment:
you should be able to test it with `io.trino.plugin.hive.optimizer.TestHivePlans#testSubsumePartitionFilterNotConvertibleToTupleDomain`
and `assertDistributedPlan`.